src/QsCompiler/Core/TransformationOptions.fs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.Core
+
+
+/// Used to configure the behavior of the default implementations for transformations.
+type TransformationOptions = {
+    /// Disables the transformation at the transformation root, 
+    /// meaning the transformation won't recur into leaf nodes or subnodes. 
+    Disable : bool
+    /// Indicates that the transformation is used to walk the syntax tree, but does not modify any of the nodes. 
+    /// If set to true, the nodes will not be rebuilt during the transformation. 
+    /// Setting this to true constitutes a promise that the return value of all methods will be ignored. 
+    DisableRebuild : bool
+}
+    with 
+    
+    /// Default transformation setting. 
+    /// The transformation will recur into leaf and subnodes, 
+    /// and all nodes will be rebuilt upon transformation.
+    static member Default = {
+        Disable = false 
+        DisableRebuild = false
+    }
+
+    /// Disables the transformation at the transformation root, 
+    /// meaning the transformation won't recur into leaf nodes or subnodes. 
+    static member Disabled = {
+        Disable = true
+        DisableRebuild = false
+    }
+
+    /// Indicates that the transformation is used to walk the syntax tree, but does not modify any of the nodes. 
+    /// All nodes will be traversed recursively, but the nodes will not be rebuilt. 
+    /// Setting this option constitutes a promise that the return value of all methods will be ignored. 
+    static member NoRebuild = {
+        Disable = false
+        DisableRebuild = true
+    }
+
+
+/// Tools for adapting the default implementations for transformations 
+/// based on the specified options. 
+module internal Utils = 
+    type internal INode =
```

Comment:
Could you get the same effect with Fold and Walk just being functions, and having Node in the various transformation classes being a function variable that you assign Fold or Walk to, just as you do now? You'd then call Node instead of Node.BuildOr -- which probably means it should be called BuildOr if you do it that way.

It doesn't make much of a difference, since you only ever have one instance of Fold and Walk anyway, so there's no memory or GC issue. It just seems a little more "F#-y".

Reply:
No, unfortunately not. F# is a bit particular with what can be generic. A member can be generic, or a let in a module, but not a let in a class. The interface safes me the evaluation of an additional if on each invokation of BuildOr.